### PR TITLE
bpo-40696: Fix a hang that can arise after gen.throw()

### PR DESCRIPTION
**Pull request by cjerdonek**
https://bugs.python.org/issue40696

### CODE REVIEW

File: Lib/test/test_asyncio/test_tasks.py [hunk 1]
```diff
@@ -540,6 +540,34 @@ async def run():
...
+                task = self.new_task(loop, process_exc(exc))
+                try:
+                    await task
+                except BaseException as exc:
+                    has_cycle = (exc is exc.__context__)
+                    # Prevent a hang if has_cycle is True.
+                    exc.__context__ = None
+
+        task = self.new_task(loop, run())
+        loop.run_until_complete(task)
```
Comment (1st1) [inline Lib/test/test_asyncio/test_tasks.py]: I suggest using try..finally always to make sure the loop is closed for failed tests.

File: Lib/test/test_asyncio/test_tasks.py [hunk 2]
```diff
@@ -540,6 +540,34 @@ async def run():
...
+        loop.close()
+        self.assertEqual(has_cycle, False)
```
Comment (1st1) [inline Lib/test/test_asyncio/test_tasks.py:574]:
```suggestion
        self.assertFalse(has_cycle)
```

File: Modules/_asynciomodule.c [hunk 3]
```diff
@@ -612,19 +613,25 @@ create_cancelled_error(PyObject *msg)
...
+future_set_cancelled_error(FutureObj *fut)
 {
-    PyObject *exc = create_cancelled_error(msg);
+    PyObject *exc = create_cancelled_error(fut->fut_cancel_msg);
     PyErr_SetObject(asyncio_CancelledError, exc);
     Py_DECREF(exc);
+
+    PyThreadState *tstate = _PyThreadState_GET();
+
+    _PyErr_StackItem *saved_exc_info = tstate->exc_info;
```
Comment (1st1) [inline Modules/_asynciomodule.c]: Oh, this is fairly invasive. Ideally we shouldn't be tinkering with the thread state in non-core files.

File: Objects/genobject.c [hunk 4]
```diff
@@ -203,13 +203,15 @@ gen_send_ex(PyGenObject *gen, PyObject *arg, int exc, int closing)
...
 
-    if (exc) {
-        _PyErr_ChainStackItem(&gen->gi_exc_state);
-    }
-
     gen->gi_running = 1;
     gen->gi_exc_state.previous_item = tstate->exc_info;
     tstate->exc_info = &gen->gi_exc_state;
+
+    /* XXX: can we assert that _PyErr_Occurred() is true when exc is true? */
```
Comment (1st1) [inline Objects/genobject.c]: I'd add an assertion and if the tests pass merge it. We can relax the check if any regression is found during the 3.9 release process--in which case we'd know the exact scenario and would be able to test it.

File: Modules/_asynciomodule.c (hunk 3 above)
Comment (1st1) [reply Modules/_asynciomodule.c]: Can we perhaps make `_PyErr_ChainThreadState` a bit more versatile to handle this use case? Maybe add a second argument?

**Review by 1st1 — COMMENTED** [review]
I've left a few comments, Chris.

File: Modules/_asynciomodule.c (hunk 3 above)
Comment (cjerdonek) [reply Modules/_asynciomodule.c]: @1st1 Thanks for the helpful comments, Yury.

> Can we perhaps make _PyErr_ChainThreadState a bit more versatile to handle this use case?

Sure, I'll propose something. This is something that had crossed my mind, too.

**Comment by cjerdonek** [comment]
@1st1 I made some changes. Let me know what you think. Happy to iterate further.

**Review by 1st1 — APPROVED** [review]

**Pull request merged**